pyramid_swagger/api.py

```diff
@@ -101,26 +101,206 @@ def view_for_api_declaration(request):
+    target_scheme = 'file' if target_url.scheme == '' else target_url.scheme
+
+    if target_scheme == 'file':
+        spec_dir = os.path.dirname(urlparse(spec.origin_url).path)
+        target_url = urlparse('{path}#{fragment}'.format(
+            path=_relpath(target_url.path, spec_dir),
+            fragment=target_url.fragment,
+        ))
+
+    if target_scheme in ['file', 'http', 'https']:
+        return target_url
+    else:
+        raise ValueError(
+            'Invalid target: {target}'.format(target=target)
+        )
+
+
+def _marshal_target(target_url):
+    target_scheme = target_url.scheme
+    if len(target_url.path) > 0 and \
+            target_scheme in ['', 'file', 'http', 'https']:
+        value = _low_level_translate(
+            '{scheme}://{host}{path}#{fragment}'.format(
+                scheme=target_url.scheme if len(target_scheme) > 0 else 'file',
+                host=target_url.hostname if target_url.hostname else '',
+                path=target_url.path,
+                fragment=target_url.fragment,
+            ))
+        return value
+    else:
+        raise ValueError(
+            'Invalid target: {target}'.format(target=urlunparse(target_url))
+        )
+
+
+def _unmarshal_target(marshaled_target):
+    if any(marshaled_target.startswith(x) for x in
+           ('file.', 'http.', 'https.')):
+        result = _low_level_translate(marshaled_target, is_marshaling=False)
+        # Remove the file:// scheme to allow the use of relative paths
+        if not result.startswith('file:///'):
+            result = result.replace('file://', '')
+        return result
+    else:
+        raise ValueError(
+            'Invalid marshaled object: {target}'.format(
+                target=marshaled_target
+            )
+        )
+
+
+def is_a_definition(json_path):
+    # definition is possible if:
+    # json_path = ['definitions']
+    # json_path = [ ..., 'schema']
+    # json_path = [ ..., 'items']
+    # json_path = [ ..., 'allOf']
+    return json_path == ['definitions'] or \
+           json_path[-1] in ['schema', 'items', 'allOf']
+
+
+def resolve_ref(spec, url, json_path, file_path, defs_dict):
     with spec.resolver.resolving(url) as resolved_spec_dict:
         spec_dict = copy.deepcopy(resolved_spec_dict)
-        return resolve_refs(spec, spec_dict)
+        return resolve_refs(spec, spec_dict, json_path, file_path, defs_dict)
 
 
-def resolve_refs(spec, val):
+def resolve_refs(spec, val, json_path, file_path, defs_dict):
     if isinstance(val, dict):
         new_dict = {}
         for key, subval in val.items():
             if key == '$ref':
+                if is_a_definition(json_path):
+                    target_url = _get_target(spec, subval, file_path)
+                    marshaled_target = _marshal_target(target_url)
+                    if marshaled_target not in defs_dict:
+                        defs_dict[marshaled_target] = None
```

Comment:
Maybe adding a comment here would have been beneficial 😄 
The placeholder (``None``) is needed because the ``resolve_ref`` could involve a recursive method call if there is a recursive data-model dereferencing and we need to have a stop condition for the recursion.

Since we are resolving the references